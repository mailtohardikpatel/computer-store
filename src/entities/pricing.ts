type RuleType = "payForLess" | "discountedPrice";

interface RuleDefinition {
  quantity?: number;
  discountedQuantity?: number;
  bulkPrice?: number;
}

export class PricingRule {
  constructor(
    public name: string,
    public detail: string,
    public skuId: string,
    public type: RuleType,
    public definition: RuleDefinition
  ) {}

  apply(quantity: number, price: number): number {
    switch (this.type) {
      case "payForLess":
        return this.calculatePayForLessPrice(quantity, price);

      case "discountedPrice":
        return this.calculateDiscountedPrice(quantity, price);

      default:
        return quantity * price;
    }
  }

  private calculatePayForLessPrice(quantity: number, price: number): number {
    const discountSets = Math.floor(quantity / this.definition.quantity!);
    const remainingItems = quantity % this.definition.quantity!;
    return (
      (discountSets * this.definition.discountedQuantity! + remainingItems) *
      price
    );
  }

  private calculateDiscountedPrice(quantity: number, price: number): number {
    if (quantity >= this.definition.quantity!) {
      return quantity * this.definition.bulkPrice!;
    }
    return quantity * price;
  }
}