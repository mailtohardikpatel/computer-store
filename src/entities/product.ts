export class Product {
    constructor(public sku: string, public name: string, public price: number){}
}