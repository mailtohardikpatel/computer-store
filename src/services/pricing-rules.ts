import { PricingRule } from "../entities/pricing";

export class PricingRules {
  private rules: Map<string, PricingRule>;

  constructor(rules: PricingRule[] = []) {
    this.rules = new Map();
    rules.forEach((rule) => this.rules.set(rule.skuId, rule));
  }

  getRule(skuId: string): PricingRule | undefined {
    return this.rules.get(skuId);
  }

  getAllRules(): PricingRule[] {
    return Array.from(this.rules.values());
  }
}