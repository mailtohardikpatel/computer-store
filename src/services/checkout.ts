import products from "../data/products.json"
import { PricingRules } from "./pricing-rules";

export class Checkout {
  private itemCounts: Map<string, number>;

  constructor(private pricingRules: PricingRules) {
    this.itemCounts = new Map();
  }

  scan(skuId: string): void {
    this.itemCounts.set(skuId, (this.itemCounts.get(skuId) || 0) + 1);
  }

  total(): string {
    let total = 0;
    for (const [skuId, count] of this.itemCounts.entries()) {
        const price = products.find( p => p.id === skuId )?.price!
        const rule = this.pricingRules.getRule(skuId);

        if (rule) {
            total += rule.apply(count, price);
        } else {
            total += count * price;
        }
    }
    return total.toFixed(2);
  }
}