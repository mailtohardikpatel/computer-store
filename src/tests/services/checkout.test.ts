import { PricingRule } from "../../entities/pricing";
import { Checkout } from "../../services/checkout";
import { PricingRules } from "../../services/pricing-rules";

// Mock data for products
const mockInventoryData = JSON.stringify([
  { id: "atv", name: "Apple TV", price: 109.5 },
  { id: "ipd", name: "Super iPad", price: 549.99 },
  { id: "mbp", name: "MacBook Pro", price: 1399.99 },
  { id: "vga", name: "VGA adapter", price: 30.0 },
]);

describe("Checkout", () => {
  let checkout: Checkout;
  const pricingRulesList = new PricingRules([
    new PricingRule("3for2deal","This is for Apple TVs.","atv", "payForLess", { quantity: 3, discountedQuantity: 2 }),
    new PricingRule("bulkDiscount","This is for iPad.","ipd", "discountedPrice", { quantity: 4, bulkPrice: 499.99 }),
  ]);

  beforeEach(() => {
    checkout = new Checkout(pricingRulesList);
  });
  test("should calculate total with no items", () => {
    const total = checkout.total();
    expect(total).toBe("0.00");
  });

  test("should calculate total with scanned items", () => {
    checkout.scan("ipd");
    checkout.scan("atv");
    const total = checkout.total();
    expect(total).toMatchSnapshot();
  });

  test("should apply discount correctly for scanned items", () => {
    checkout.scan("atv");
    checkout.scan("atv");
    checkout.scan("atv");
    const total = checkout.total();
    expect(total).toMatchSnapshot();
  });
})