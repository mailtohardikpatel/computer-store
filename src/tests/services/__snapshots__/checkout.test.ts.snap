// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkout should apply discount correctly for scanned items 1`] = `"219.00"`;

exports[`Checkout should calculate total with scanned items 1`] = `"659.49"`;
